Allow custom compile options for StanModel

#### Summary:
When debugging a model, I care more about compile time than speed.  I would like to be able to specify a -O0 build (or other compile options) when creating a StanModel.


#### Description:
The compile flags in StanModel are hardcoded in extra_compile_args are hardcoded.  I would like a Python keyword argument to override these defaults, say StanModel(code=code, compile_flags=['-O0']).  I would suggest that these new flags are appended to the end of extra_compile_flags.  Also, I am frequently using Stan within an IPython Notebook, so environment variables would be more annoying to use than an extra keyword argument.


#### Reproducible Steps:
N/A


#### Current Output:
StanModel always compiles with -O2.


#### Expected Output:
StanModel can compile with -O0.


#### PyStan Version:
2.9.0.0

#### Python Version:
2.7

#### Operating System:
Ubuntu Linux

Thank you.

Oh, one more thing related to the compile flags (sorry about the slow drip of information).  My compiler is emitting a couple of other errors that are unimportant but distracting.  It seems like -Wall is somewhere in your compilation flags, but any warnings from C++ hardly seem useful to a user.  It is unlikely that C++ warnings from the generated code will be useful in diagnosing a problem in the Stan model.

cc1plus: warning: command line option â-Wstrict-prototypesâ is valid for C/ObjC but not for C++
In file included from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1781:0,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:309:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^
In file included from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/multi_array/base.hpp:28:0,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/multi_array.hpp:21,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/numeric/odeint/util/multi_array_adaption.hpp:29,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/numeric/odeint.hpp:61,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/prim/arr/functor/integrate_ode.hpp:12,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math.hpp:65,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/src/stan/variational/advi.hpp:4,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan_fit.hpp:83,
                 from /tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:313:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/multi_array/concept_checks.hpp: In static member function âstatic void boost::multi_array_concepts::detail::idgen_helper<N>::call(Array&, const IdxGen&, Call_Type)â:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/multi_array/concept_checks.hpp:42:43: warning: typedef âindex_rangeâ locally defined but not used [-Wunused-local-typedefs]
       typedef typename Array::index_range index_range;
                                           ^
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/multi_array/concept_checks.hpp:43:37: warning: typedef âindexâ locally defined but not used [-Wunused-local-typedefs]
       typedef typename Array::index index;
efs]
       typedef typename Array::index_range index_range;
                                           ^
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/lib/boost_1.58.0/boost/multi_array/concept_checks.hpp:54:37: warning: typedef âindexâ locally defined but not used [-Wunused-local-typedefs]
       typedef typename Array::index index;
                                     ^
In file included from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math.hpp:80:0,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/src/stan/variational/advi.hpp:4,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan_fit.hpp:83,
                 from /tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:313:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/prim/mat/err/check_positive_ordered.hpp: In function âbool stan::math::check_positive_ordered(const char*, const char*, const Eigen::Matrix<Scalar, -1, 1>&)â:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/prim/mat/err/check_positive_ordered.hpp:39:67: warning: typedef âsize_typeâ locally defined but not used [-Wunused-local-typedefs]
       typedef typename index_type<Matrix<T_y, Dynamic, 1> >::type size_type;
                                                                   ^
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp: In function âPyObject* __pyx_pf_69stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562_2_call_sampler(PyObject*, PyObject*, PyObject*, PyObject*)â:
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:8074:30: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     __pyx_t_12 = ((__pyx_t_9 != __pyx_v_fitptr->param_names_oi().size()) != 0);
                              ^
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp: In function âstd::vector<long unsigned int> __pyx_convert_vector_from_py_size_t(PyObject*)â:
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:19141:77: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     __pyx_t_5 = __Pyx_PyInt_As_size_t(__pyx_v_item); if (unlikely(__pyx_t_5 == -1 && PyErr_Occurred())) {__pyx_filen
                                                                             ^
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:506:43: note: in definition of macro âunlikelyâ
   #define unlikely(x) __builtin_expect(!!(x), 0)
                                           ^
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp: In function âstd::vector<unsigned int> __pyx_convert_vector_from_py_unsigned_int(PyObject*)â:
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:19738:83: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     __pyx_t_5 = __Pyx_PyInt_As_unsigned_int(__pyx_v_item); if (unlikely(__pyx_t_5 == -1 && PyErr_Occurred())) {__pyx
                                                                                   ^
/tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:506:43: note: in definition of macro âunlikelyâ
   #define unlikely(x) __builtin_expect(!!(x), 0)
                                           ^
In file included from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/rev/core/operator_unary_plus.hpp:7:0,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/rev/core.hpp:34,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/rev/mat/fun/Eigen_NumTraits.hpp:4,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/rev/mat/fun/grad.hpp:6,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/src/stan/model/util.hpp:7,
                 from /home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan_fit.hpp:18,
                 from /tmp/tmp6c_iMJ/stanfit4knowledge_74b131152b913b180931446518e4b25a_813617668659442562.cpp:313:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/prim/scal/fun/constants.hpp: At global scope:
/home/jumper/notebook_venv/venv/local/lib/python2.7/site-packages/pystan/stan/lib/stan_math/stan/math/prim/scal/fun/constants.hpp:66:18: warning: âstan::math::NEGATIVE_EPSILONâ defined but not used [-Wunused-variable]
     const double NEGATIVE_EPSILON

--- a/pystan/model.py
+++ b/pystan/model.py
@@ -200,7 +200,8 @@ class StanModel:
     """
     def __init__(self, file=None, charset='utf-8', model_name="anon_model",
                  model_code=None, stanc_ret=None, boost_lib=None,
-                 eigen_lib=None, verbose=False, obfuscate_model_name=True):
+                 eigen_lib=None, verbose=False, obfuscate_model_name=True,
+                 extra_compile_args=None):
 
         if stanc_ret is None:
             stanc_ret = pystan.api.stanc(file=file,
@@ -274,18 +275,15 @@ class StanModel:
         ]
         # compile stan models with optimization (-O2)
         # (stanc is compiled without optimization (-O0) currently, see #33)
-        extra_compile_args = [
-            '-O2',
-            '-ftemplate-depth-256',
-            '-Wno-unused-function',
-            '-Wno-uninitialized',
-        ]
-
-        if platform.platform().startswith('Win'):
+        if extra_compile_args is None:
             extra_compile_args = [
-                '/EHsc',
-                '-DBOOST_DATE_TIME_NO_LIB',
+                '-O2',
+                '-ftemplate-depth-256',
+                '-Wno-unused-function',
+                '-Wno-uninitialized',
             ]
+            if platform.platform().startswith('Win'):
+                extra_compile_args = ['/EHsc', '-DBOOST_DATE_TIME_NO_LIB']
 
         distutils.log.set_verbosity(verbose)
         extension = Extension(name=self.module_name,

--- a/pystan/tests/test_extra_compile_args.py
+++ b/pystan/tests/test_extra_compile_args.py
@@ -1,0 +1,21 @@
+import unittest
+
+import pystan
+
+
+class TestExtraCompileArgs(unittest.TestCase):
+
+    def test_extra_compile_args(self):
+        extra_compile_args = [
+            '-O3',
+            '-ftemplate-depth-1024',
+            '-Wno-unused-function',
+            '-Wno-uninitialized',
+        ]
+        model_code = 'parameters {real y;} model {y ~ normal(0,1);}'
+        model = pystan.StanModel(model_code=model_code, model_name="normal1",
+                                 extra_compile_args=extra_compile_args)
+        fit = model.sampling()
+        extr = fit.extract()
+        y_last, log_prob_last = extr['y'][-1], extr['lp__'][-1]
+        self.assertEqual(fit.log_prob(y_last), log_prob_last)

